control-plane/connect-inject/controllers/endpointsv2/endpoints_controller.go

```diff
@@ -0,0 +1,351 @@
+		return ctrl.Result{}, nil
+	}
+
+	// Create Consul resource service client for this reconcile.
+	resourceClient, err := consul.NewResourceServiceClient(r.ConsulServerConnMgr)
+	if err != nil {
+		r.Log.Error(err, "failed to create Consul resource client", "name", req.Name, "ns", req.Namespace)
+		return ctrl.Result{}, err
+	}
+
+	// If the Endpoints object has been deleted (and we get an IsNotFound error),
+	// we need to deregister that service from Consul.
+	err = r.Client.Get(ctx, req.NamespacedName, &endpoints)
+	if k8serrors.IsNotFound(err) {
+		err = r.deregisterService(ctx, resourceClient, req.Name, r.getConsulNamespace(req.Namespace), r.getConsulPartition())
+		return ctrl.Result{}, err
+	} else if err != nil {
+		r.Log.Error(err, "failed to get Endpoints", "name", req.Name, "ns", req.Namespace)
+		return ctrl.Result{}, err
+	}
+	r.Log.Info("retrieved Endpoints", "name", req.Name, "ns", req.Namespace)
+
+	// We expect this to succeed if the Endpoints fetch for the Service succeeded.
+	err = r.Client.Get(r.Context, types.NamespacedName{Name: endpoints.Name, Namespace: endpoints.Namespace}, &service)
+	if err != nil {
+		r.Log.Error(err, "failed to get Service", "name", req.Name, "ns", req.Namespace)
+		return ctrl.Result{}, err
+	}
+	r.Log.Info("retrieved Service", "name", req.Name, "ns", req.Namespace)
+
+	// Determine Consul service selector values from pod names.
+	podPrefixes := make(map[string]any)
+	podExactNames := make(map[string]any)
+	for _, subset := range endpoints.Subsets {
+		for address := range allAddresses(subset) {
+			if address.TargetRef != nil && address.TargetRef.Kind == "Pod" {
+				podName := types.NamespacedName{Name: address.TargetRef.Name, Namespace: req.Namespace}
+
+				// Accumulate owner prefixes and exact pod names for Consul workload selector.
+				// If this pod is already covered by a known owner prefix, skip it.
+				// If not, fetch the owner. If the owner has a unique prefix, add it to known prefixes.
+				// If not, add the pod name to exact name matches.
+				maybePodOwnerPrefix := getOwnerPrefixFromPodName(podName.Name)
+				if _, ok := podPrefixes[maybePodOwnerPrefix]; !ok {
+					var pod corev1.Pod
+					if err = r.Client.Get(ctx, podName, &pod); err != nil {
+						r.Log.Error(err, "failed to get pod", "name", podName.Name, "ns", req.Namespace)
+						errs = multierror.Append(errs, err)
+						continue
+					}
+					// Add to workload selector values.
+					// Pods can appear more than once in Endpoints subsets, so we use a set for exact names as well.
+					if prefix := getOwnerPrefixFromPod(&pod); prefix != "" {
+						podPrefixes[prefix] = true
+					} else {
+						podExactNames[podName.Name] = true
+					}
+				}
+			}
+		}
+	}
+
+	//TODO(zalimeni) Do we need to check service-enable label here on service/deployments/replicasets?
+	// Should we expect owner is annotated, and if so, is that consistent for non-replicaset workloads?
+	// If so, update to register and deregister conditionally.
```

Comment:
[This](https://docs.google.com/document/d/1q01OHEnQ6Ir40y5GeVq53X5iN-5YExHshX_pashLvlQ/edit?disco=AAAA2PmcWks) 🧵 discusses maybe not needing explicit exclusion for services. I'm not 100% certain whether this controller should expect any opt-outs that will impact service registration, outside of disabled k8s namespaces from config, already checked above.

Input welcome! Lacking a strong opinion, I'll probably convert this to a follow-up ticket as a placeholder for a pre-GA decision.